### ISSUE TITLE
Sporadic DNS timeouts on AKS w/ Cilium and Azure Firewall

### ISSUE THREAD

**Issue by EppO** [cilium/cilium#25494]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Using [AKS with cilium](https://learn.microsoft.com/en-us/azure/aks/azure-cni-powered-by-cilium) and using an Azure Firewall as my upstream DNS server (acting as proxy to Azure DNS servers), CoreDNS can't reach the private IP of the Azure Firewall DNS port sometimes. It looks quite random, here is a graph showing the occurrences over 7 days:

![Screenshot 2023-05-16 at 5 25 41 PM](https://github.com/cilium/cilium/assets/6111/581b9e50-0411-4dfb-9a41-9a26a48af02e)

CoreDNS logs looks like this (DNS requests can be anything a pod needs in my cluster, it's not tied to a particular hostname) 
```
[ERROR] plugin/errors: 2 global.handler.control.monitor.azure.com. A: read udp 10.42.0.42:55893->10.161.1.4:53: i/o timeout
```

Besides these sporadic and erratic CoreDNS errors, I didn't see anything relevant in Cilium logs. Cilium Health looks fine

```
   /¯¯\
 /¯¯\__/¯¯\    Cilium:         OK
 \__/¯¯\__/    Operator:       OK
 /¯¯\__/¯¯\    Hubble:         disabled
 \__/¯¯\__/    ClusterMesh:    disabled
    \__/

DaemonSet         cilium             Desired: 4, Ready: 4/4, Available: 4/4
Deployment        cilium-operator    Desired: 2, Ready: 2/2, Available: 2/2
Containers:       cilium             Running: 4
                  cilium-operator    Running: 2
Cluster Pods:     83/83 managed by Cilium
Image versions    cilium             mcr.microsoft.com/oss/cilium/cilium:1.12.5.1: 4
                  cilium-operator    mcr.microsoft.com/oss/cilium/operator-generic:1.12.5: 2
```

I tried in different AKS version (1.24 and 1.26) and different Azure regions as well (central US and East US2), same behavior in all of them (but different timing and frequencies)

### Cilium Version

1.12.5.1

### Kernel Version

Happens in Ubuntu 22.04.2 LTS   5.15.0-1036-azure   containerd://1.6.18+azure-1
and Ubuntu 18.04.6 LTS   5.4.0-1107-azure   containerd://1.6.18+azure-1

### Kubernetes Version

Happens in v1.26.3 and v1.24.10

### Sysdump

can provide it offline

### Anything else?

I saw some other issues about dns timeouts but it was related to cilium BYO CNI and ubuntu 22.04 systemd issue, in my case it also affect Ubuntu 18.04 and I disabled the unattended upgrade of Ubuntu (to prevent systemd shenanigans). 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by phealy** [comment]
@EppO I'm from the AKS product group - can you please open a support request on this issue via Azure support?

**Comment by squeed** [comment]
These sort of wild-goose chases can be hard to narrow down. Would you be able to capture Hubble logs from a failure?

**Comment by EppO** [comment]
Unfortunately, AKS with Azure CNI Powered by Cilium doesn't support (yet??) Hubble. I'll open a case on Azure as it seems it is GA now (missed the announcement but docs removed the "preview" flag. Will keep this ticket open and gives Azure support findings

**Comment by davem-git** [comment]
I have this problem in general on my clusters without cilium on them. We're still trying to track it down.

**Comment by unix68** [comment]
seeing also same strange issues with our Azure firewall DNS proxy ... still investigating ... not sure whether it's our setup with dns forwarding server as upstream dns or whether it's the firewall going crazy sometimes / in our case it's not AKS specific

**Comment by EppO** [comment]
In my case, I ruled out the Azure firewall DNS proxy with a script that I ran outside of the AKS cluster (but in the same VNET), doing some DNS request to the same upstream CoreDNS is using (the Azure firewall DNS proxy), and there was no timeouts from the VM ever, only seeing them from the AKS cluster.

**Comment by ti-mo** [comment]
@EppO Are you using Cilium's DNS proxy, e.g. toFQDN rules? Have you run the same script from a Cilium-managed pod hitting the Azure Firewall DNS server directly?

**Comment by unix68** [comment]
> seeing also same strange issues with our Azure firewall DNS proxy ... still investigating ... not sure whether it's our setup with dns forwarding server as upstream dns or whether it's the firewall going crazy sometimes / in our case it's not AKS specific

Update: MS support ticket solved the issue / updated release of Azure FW resolved the issue

**Comment by EppO** [comment]
> @EppO Are you using Cilium's DNS proxy, e.g. toFQDN rules? Have you run the same script from a Cilium-managed pod hitting the Azure Firewall DNS server directly?

Using [Azure CNI powered by Cilium](https://learn.microsoft.com/en-us/azure/aks/azure-cni-powered-by-cilium), there is no support for Cilium's DNS proxy unfortunately, but it might change in the future: https://github.com/Azure/AKS/issues/3450

In my case, we ran into a bug with azure-cns component (Azure IPAM that cilium is using as IPAM delegated-plugin), that made the node in NotReady state when it was restarting, but it was a connectivity loss more global that just DNS timeouts like I noticed when I opened this issue initially. Since then, Azure rolled back the azure-cns change responsible of the connectivity issue and most likely did other changes and I don't see the DNS timeouts happening anymore (haven't had one for weeks). So I don't think it was cilium related after all, more likely AKS related.
